### PR TITLE
feat: yaml sorting eslint plugin

### PR DESCRIPTION
**Pull request by mshojaei-txfusion**
### Description

This PR adds a new ESLint rule `sort-yaml-arrays` to the utils package, which allows for sorting arrays in YAML files based on specified keys. This helps maintain consistent and predictable ordering in YAML configuration files.

Key features:
- New utility function for array sorting by key
- YAML transformation with complete comment preservation 
- Support for sorting nested arrays at specific paths
- Comprehensive test coverage for all sorting scenarios and comment handling

### Drive-by changes

- Enhanced the comment preservation logic in YAML transformation
- Improved unit test coverage for YAML transformation utilities

### Related issues
Required to resolve
https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/5547

### Backward compatibility

Yes - This change is fully backward compatible. It adds a new ESLint rule that can be optionally enabled.

### Testing

Unit Tests - Comprehensive test suite covers:
- Basic array sorting functionality
- Nested array sorting
- Comment preservation
- Various edge cases

### CODE REVIEW

File: typescript/utils/src/arrays.ts [hunk 1]
```diff
@@ -20,3 +22,15 @@ export function exclude<T>(item: T, list: T[]) {
 export function randomElement<T>(list: T[]) {
   return list[randomInt(list.length)];
 }
+
+export function sortArrayByKey<T extends Record<string, any>>(
+  array: T[],
+  sortKey: string,
```
Comment (xeno097) [inline typescript/utils/src/arrays.ts:28]: Can you update this to be `keyof T` instead of a plain string? so we can have type safety there too

File: typescript/utils/src/yaml.ts [hunk 2]
```diff
@@ -15,3 +22,406 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+    });
+  });
+
+  return matchingConfig?.sortKey || null;
+}
+
+/**
+ * Sorts arrays nested within objects according to configuration
+ */
+export function sortNestedArrays<T = any>(
```
Comment (xeno097) [inline typescript/utils/src/yaml.ts:397]: Can we use [this](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/typescript/utils/src/objects.ts#L355) in here for the recursive logic?

File: typescript/utils/src/yaml.ts [hunk 3]
```diff
@@ -15,3 +22,406 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+  data: T,
+  config: ArraySortConfig,
+  path: string[] = [],
+): T {
+  // Handle primitive values
+  if (!isObject(data) && !isArray(data)) {
```
Comment (antigremlin) [inline typescript/utils/src/yaml.ts:403]: Why not reverse the conditions?

```typescript
if (Array.isArray(data)) {}
if (typeof data === 'object' && data !== null) {}
return data;
```

File: typescript/utils/src/yaml.ts [hunk 4]
```diff
@@ -15,3 +22,406 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+  if (isArray(data)) {
+    const sortKey = findSortKeyForPath(path, config);
+
+    // Process each array item recursively
+    const processedArray = data.map((item, idx) =>
+      sortNestedArrays(item, config, [...path, idx.toString()]),
+    );
+
+    return (
+      sortKey ? sortArrayByKey(processedArray, sortKey) : processedArray
```
Comment (antigremlin) [inline typescript/utils/src/yaml.ts:417]: Do we really need a dedicated function?

```typescript
!sortKey
  ? processedArray
  : processedArray.sort((a, b) => 
      a[sortKey] < b[sortKey] ? -1 : a[sortKey] > b[sortKey] ? 1 : 0
```

File: typescript/utils/src/yaml.ts [hunk 5]
```diff
@@ -15,3 +22,406 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+
+    if (path.length !== patternParts.length) {
+      return false;
+    }
+
+    return path.every((part, idx) => {
+      const patternPart = patternParts[idx];
+      if (patternPart === '*') return true;
+      if (patternPart.endsWith('[]')) return patternPart.slice(0, -2) === part;
+      return patternPart === part;
```
Comment (antigremlin) [inline typescript/utils/src/yaml.ts:387]: What is the purpose of a pattern like `foo[].bar`? Or does it work as intended
Will this handle a path like `foo.4.bar`?

It seems we may need tests for this algorithm.

File: typescript/utils/src/eslint-rules/sort-yaml-arrays.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+          if (!yamlData) return;
+
+          const sortedData = sortNestedArrays(yamlData, sortConfig);
+
+          const sortedYaml = YAML.stringify(sortedData);
+          const originalYaml = YAML.stringify(yamlData);
+
+          if (sortedYaml !== originalYaml) {
+            context.report({
+              node,
```
Comment (antigremlin) [inline typescript/utils/src/eslint-rules/sort-yaml-arrays.ts:65]: Does it report errors on a particular line of the YAML file, or always on the first line?

File: typescript/utils/src/yaml.test.ts [hunk 7]
```diff
@@ -31,3 +37,366 @@ describe('tryParseJsonOrYaml', () => {
...
+  it('should sort a simple array by key', () => {
+    const data = {
+      items: [
+        { name: 'c', value: 3 },
+        { name: 'a', value: 1 },
+        { name: 'b', value: 2 },
+      ],
+    };
+
+    const config: ArraySortConfig = {
```
Comment (antigremlin) [inline typescript/utils/src/yaml.test.ts:51]: How about some objects in the array not having the expected key?

**Review by antigremlin — COMMENTED** [review]
This is a great effort, considering all the complexity of preserving comments. It may need more work though, considering the existing test cases.

Have we considered extending the rules in `eslint-plugin-yml`? It takes a parser-driven approach.

**Comment by mshojaei-txfusion** [comment]
> This is a great effort, considering all the complexity of preserving comments. It may need more work though, considering the existing test cases.
> 
> Have we considered extending the rules in `eslint-plugin-yml`? It takes a parser-driven approach.

First tried with `yaml-eslint-parser` but found it so complex and this way was more straight forward and easier to test.

File: typescript/utils/src/arrays.ts (hunk 1 above)
Comment (mshojaei-txfusion) [reply typescript/utils/src/arrays.ts]: Done!

File: typescript/utils/src/yaml.ts (hunk 2 above)
Comment (mshojaei-txfusion) [reply typescript/utils/src/yaml.ts:397]: For this we need to apply this change there, that I think would break other functionalities

**Review by xeno097 — APPROVED** [review]
approved on my side but wait for @antigremlin to check the pending comments

File: typescript/utils/src/eslint-rules/sort-yaml-arrays.ts (hunk 6 above)
Comment (mshojaei-txfusion) [reply typescript/utils/src/eslint-rules/sort-yaml-arrays.ts:65]: No it wont mention line, do you think we need to add it?

File: typescript/utils/src/yaml.ts (hunk 4 above)
Comment (antigremlin) [reply typescript/utils/src/yaml.ts:442]: It's OK, let it stay as it is. Just for clarity, here's why I raised this:

- I agree, that having a dedicated utility function can make the code more readable for someone familiar with the codebase, especially when this function is used often.
- However, such functions often are in a different file (or even a different package). This introduces dependencies, and increases cognitive load when reading/writing code. It goes like this:

> What does it do? I don't exactly remember... Let me navigate to source and read it. Oh, it's the same as that lodash func, right...

- Making these generic utilities type-safe requires extra effort
- You wrote tests for this, even though the underlying `sortBy` is likely well-tested
- This code depends on `lodash`

I'm taking this a bit too far, maybe. But you see my point 😄

File: typescript/utils/src/eslint-rules/sort-yaml-arrays.ts (hunk 6 above)
Comment (antigremlin) [reply typescript/utils/src/eslint-rules/sort-yaml-arrays.ts:65]: No, with the current approach it would be quite difficult. I wanted to be sure I understood this.

File: typescript/utils/src/yaml.test.ts [hunk 8]
```diff
@@ -31,3 +37,366 @@ describe('tryParseJsonOrYaml', () => {
...
+          { code: 'cat-a', items: [] },
+        ],
+      },
+    };
+
+    const config: ArraySortConfig = {
+      arrays: [
+        { path: 'level1.items', sortKey: 'id' },
+        { path: 'level1.categories', sortKey: 'code' },
+        { path: 'level1.categories.*.items', sortKey: 'priority' },
```
Comment (antigremlin) [inline typescript/utils/src/yaml.test.ts:90]: It seems it works the same as `level1.categories[].items`. But we're not testing the array syntax. Do we really need both `categories[]` and `categories.*`?

File: typescript/utils/src/yaml.test.ts [hunk 9]
```diff
@@ -31,3 +37,366 @@ describe('tryParseJsonOrYaml', () => {
...
+            { id: 'y', completed: false },
+          ],
+        },
+      ],
+    };
+
+    const config: ArraySortConfig = {
+      arrays: [
+        { path: 'users', sortKey: 'name' },
+        { path: 'users.*.tasks', sortKey: 'id' },
```
Comment (antigremlin) [inline typescript/utils/src/yaml.test.ts:140]: Same comment about `users[]` vs. `users.*`

File: typescript/utils/src/yaml.ts [hunk 10]
```diff
@@ -15,3 +21,436 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+    pathIndex++;
+  }
+
+  return pathIndex === path.length;
+}
+
+/**
+ * Sorts arrays nested within objects according to configuration
+ */
+export function sortNestedArrays<T = any>(
```
Comment (antigremlin) [inline typescript/utils/src/yaml.ts:427]: It seems the type param doesn't need the `= any` part.

**Review by antigremlin — APPROVED** [review]
Nice!

File: typescript/utils/src/arrays.ts [hunk 11]
```diff
@@ -20,3 +22,15 @@ export function exclude<T>(item: T, list: T[]) {
...
+
+export function sortArrayByKey<T extends Record<keyof T, any>>(
+  array: T[],
+  sortKey: keyof T,
+): T[] {
+  return sortBy(array, [
+    (item) => {
+      const value = item[sortKey];
+      return value;
+    },
```
[the remark this answers is not in the document]
Comment (antigremlin) [reply typescript/utils/src/arrays.ts:34]: 👎 Incorrect. `_.sortBy()` appears to put undefined and null values at the end of the sort.

File: typescript/utils/src/arrays.ts [hunk 12]
```diff
@@ -20,3 +22,15 @@ export function exclude<T>(item: T, list: T[]) {
 export function randomElement<T>(list: T[]) {
   return list[randomInt(list.length)];
 }
+
+export function sortArrayByKey<T extends Record<keyof T, any>>(
+  array: T[],
+  sortKey: keyof T,
+): T[] {
+  return sortBy(array, [
```
Comment (antigremlin) [inline typescript/utils/src/arrays.ts:30]: Why not `return sortBy(array, [(item) => item[sortKey]);` ?

File: typescript/utils/src/yaml.test.ts (hunk 8 above)
Comment (mshojaei-txfusion) [reply typescript/utils/src/yaml.test.ts:90]: categories[] is for arrays, while categories.* allows any object or array under a category, but the next level matters. So, `*` works for both, but `[]` covers cases where we might not intend to target objects.

Comment (mshojaei-txfusion) [reply typescript/utils/src/yaml.test.ts:90]: Anyway will improve tests to cover this edge case

File: typescript/utils/src/yaml.ts [hunk 13]
```diff
@@ -15,3 +21,436 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+      pathIndex++;
+      continue;
+    }
+
+    if (pattern !== pathSegment) return false;
+    pathIndex++;
+  }
+
+  return pathIndex === path.length;
+}
```
[the remark this answers is not in the document]
Comment (antigremlin) [reply typescript/utils/src/yaml.ts:422]: @mshojaei-txfusion WDYT?

**Review by antigremlin — COMMENTED** [review]
Please have a look at the remaining comments from the Graphite bot.

**Review by ltyu — APPROVED** [review]

File: typescript/utils/src/yaml.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (mortezashojaei) [reply typescript/utils/src/yaml.ts:422]: Anyway added this test case to cover this scenario: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/5909/commits/1b49b9486af1bc3245d5e379a1954f1f4f5a0b54

File: typescript/utils/src/yaml.ts [hunk 14]
```diff
@@ -15,3 +21,436 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
...
+      const prefix = pattern.slice(0, -2);
+      if (!prefix || pathSegment !== prefix) return false;
+
+      pathIndex++;
+      if (pathIndex >= path.length) return false;
+
+      if (isNaN(Number(path[pathIndex]))) return false;
+      pathIndex++;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (mortezashojaei) [reply typescript/utils/src/yaml.ts:415]: We dont want to match in case `anything.[].anything` as we already have * for this case!

File: typescript/utils/src/eslint-rules/sort-yaml-arrays.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            });
+          }
+        } catch (error: unknown) {
+          context.report({
+            node,
+            message: `Error processing YAML: ${
+              error instanceof Error ? error.message : String(error)
+            }`,
+          });
+        }
```
[the remark this answers is not in the document]
Comment (mortezashojaei) [reply typescript/utils/src/eslint-rules/sort-yaml-arrays.ts:83]: No need as this is already part of a return:

```
    return {
      Program(node: Rule.Node) {
```

File: typescript/utils/src/eslint-rules/sort-yaml-arrays.test.ts [hunk 16]
```diff
@@ -0,0 +1,647 @@
...
+      expected: `features:
+  - name: search
+    enabled: false
+  - name: auth
+    enabled: true
+  - name: notifications
+    enabled: true`,
+      options: {
+        arrays: [{ path: 'features', sortKey: 'enabled' }],
+      },
```
[the remark this answers is not in the document]
Comment (mortezashojaei) [reply typescript/utils/src/eslint-rules/sort-yaml-arrays.test.ts]: Good catch, fixed

**Review by antigremlin — APPROVED** [review]
